consult-selectrum.el

```diff
@@ -45,11 +45,7 @@
 (defun consult-selectrum--refresh ()
   "Refresh selectrum view."
   (when (eq completing-read-function #'selectrum-completing-read)
-    ;; TODO async refresh hack!
-    (let ((old-index selectrum--current-candidate-index))
-      (selectrum-exhibit)
-      (setq selectrum--current-candidate-index old-index)
-      (selectrum--minibuffer-post-command-hook))))
+    (selectrum-exhibit 'keep-selected)))
```

Comment:
I'll take a stab at writing the icomplete version of this. It will have to be slightly different because icomplete does not always have a candidate selected. Before you cycle among the candidates, none of the them is really selected. For example pressing `RET` in that circumstance does not enter the top candidate (unless it is the only one). But if you run `icomplete-forward-completions` and then immediately `icomplete-backward-completions` everything looks the same as if you hadn't cycled, the minibuffer input doesn't change, but now if you press `RET` you do get the top candidate.

Given this, I believe the correct behavior for icomplete when updating the candidate list is:

- if no candidate was selected yet, stick to no candidate being selected
- if a candidate was selected and it happens to still exist after the update, select it again
- if a candidate was selected but it disappears during the update, go to the no candidate selected state.

Reply:
Does that sound reasonable?

Reply:
This sounds great! Thank you so much for taking a stab at this! I am perfectly happy with having different behaviors for different backends, since this should match the behavior expected by the user of the particular completion system!
@clemera [already wrote the version for Selectrum](https://github.com/raxod502/selectrum/pull/307), which achieves option 3 from above! It worked well in my tests. I am not yet 100% happy with the patch, but I am sure we will see this being merged very soon!

Reply:
Mmh. I'm wrong about the behavior of icomplete. It's not really that sometimes no candidate is selected. I would say the following is more accurate:

- The first candidate is always the "selected one", in the sense that `icomplete-force-complete-and-exit` selects that one (whether or not you have been cycling).
- The default way to exit icomplete is `icomplete-ret`, which does choose the first candidate if you have not entered any text into the minibuffer (more precisely if the part between the completion boundaries is empty, so for example if in find-file you navigate to some directory, the input is considered "empty" if you haven't type anything for the last path component yet), but will not choose the first candidate if you have entered some text into the minibuffer.

Given that corrected account I believe the correct behavior for refreshing icomplete is to search for the previous first candidate in the new candidate list and if it does appear, rotate the new candidate list to make that the first candidate, and if it does not appear, do not rotate the new candidate list.

Reply:
Oops, I ran into trouble @minad. I'm trying to test my icomplete refresh function and I wanted to use a very slowly updating async source, so I wrote this little script in a file called `slowsource` with I made executable and put on the `$PATH`:

```perl
#! /usr/bin/perl

for (1 .. 10000) {
    print int(rand(1000000)), "\n";
    sleep 1;
}
```

and this function to use it:

```elisp
(defun test-async ()
  (interactive)
  (consult--read
   "Number: "
   (thread-first (consult--async-sink)
     (consult--async-refresh)
     (consult--async-indicator)
     (consult--async-process '("slowsource")))))
```

But I think I must have misunderstood the API because I don't see any candidates appearing. What am I doing wrong?

Reply:
There is nothing wrong with your usage of the api. What is missing is the flush in your perl. In case there is another error, there is also the async stderr ephemeral buffer.

Reply:
For testing it may be easier to use timer based source? The async is just a function and you can push candidates from anywhere.

~~~
(funcall async '(list of candidates))
~~~

Reply:
Thank you! Classic trap, the buffering. Sorry to bother you for something silly.

Reply:
No worries! Thank you again for looking into this - I wouldn't be able to do this quickly without your help, since I am not familiar with the icomplete internals. Btw, how would this thing work in combination with embark-live-occur? If I am entering something, live-occur updates as does the standard emacs TAB completion. But can I also automatically refresh the live-occur somehow?

Reply:
For Embark live occur its easier since there is no notion of the selected candidate, so this suffices:

```elisp
(add-hook 'consult--completion-refresh-hook #'embark-occur--update-linked)
```

Reply:
OK, done! I made PR against the async branch.

Reply:
Thanks!